abstract-iterator.js

```diff
@@ -271,46 +261,48 @@ class CommonIterator {
     })
   }
 
-  close (callback) {
-    callback = fromCallback(callback, kPromise)
-
+  async close () {
     if (this[kClosed]) {
-      this.nextTick(callback)
-    } else if (this[kClosing]) {
-      this[kCloseCallbacks].push(callback)
+      return
+    }
+
+    if (this[kClosing]) {
+      // First caller of close() is responsible for error
+      return this[kClosingPromise].catch(noop)
     } else {
       this[kClosing] = true
-      this[kCloseCallbacks].push(callback)
 
-      if (!this[kWorking]) {
-        this._close(this[kHandleClose])
-      } else if (this[kAbortOnClose]) {
-        // Don't wait for work to finish. Subsequently ignore the result.
-        const cb = this[kFinishWork]()
+      if (this[kWorking]) {
+        // Wait for work, but handle closing and its error here.
+        this[kClosingPromise] = new Promise((resolve, reject) => {
+          this[kPendingClose] = () => {
+            this[kCallClose]().then(resolve, reject)
+          }
+        })
 
-        cb(new ModuleError('Aborted on iterator close()', {
-          code: 'LEVEL_ITERATOR_NOT_OPEN'
-        }))
+        // If implementation supports it, abort the work.
+        this[kAbortController].abort()
+      } else {
+        this[kClosingPromise] = this[kCallClose]()
       }
-    }
-
-    return callback[kPromise]
-  }
 
-  _close (callback) {
-    this.nextTick(callback)
+      return this[kClosingPromise]
+    }
   }
 
-  [kHandleClose] () {
-    this[kClosed] = true
-    this.db.detachResource(this)
+  async _close () {}
 
-    const callbacks = this[kCloseCallbacks]
-    this[kCloseCallbacks] = []
+  async [kCallClose] () {
+    this[kPendingClose] = null
 
-    for (const cb of callbacks) {
-      cb()
+    try {
+      await this._close()
+    } catch (err) {
+      this[kClosed] = true
```

Comment:
Should also be set if there was no error. Not a real bug though, just means an extra tick is spent in the async iterator when it checks `if (!this[kClosed]) await this.close()`.